cpp/src/join/full_join.cuh

```diff
@@ -0,0 +1,181 @@
+        const size_type buffer_size,
+        const size_type expand_size) {
+    size_type requested_size = buffer_size + expand_size;
+    //No need to proceed if the buffer can contain requested additional elements
+    if (*buffer_capacity >= requested_size) {
+        return GDF_SUCCESS;
+    }
+    data_type * new_buffer{nullptr};
+    data_type * old_buffer = *buffer;
+    CUDA_TRY( cudaMalloc(&new_buffer, requested_size*sizeof(data_type)) );
+    CUDA_TRY( cudaMemcpy(new_buffer, old_buffer, buffer_size*sizeof(data_type), cudaMemcpyDeviceToDevice) );
+    CUDA_TRY( cudaFree(old_buffer) );
+    *buffer = new_buffer;
+    *buffer_capacity = requested_size;
+
+    return GDF_SUCCESS;
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Adds indices that are missing in r_index_ptr at the ends and places
+* JoinNoneValue to the corresponding l_index_ptr.
+*
+* @Param l_index_ptr Address of the left indices
+* @Param r_index_ptr Address of the right indices
+* @Param index_capacity Amount of memory allocated for left and right indices
+* @Param index_size Number of left and right indices
+* @Param max_index_value The maximum value an index can have in r_index_ptr
+* @tparam index_type The type of data associated with index_ptr
+* @tparam size_type The data type used for size calculations
+*
+* @Returns  cudaSuccess upon successful completion of append call. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename index_type, typename size_type>
+gdf_error append_full_join_indices(
+        index_type ** l_index_ptr,
+        index_type ** r_index_ptr,
+        size_type * const index_capacity,
+        size_type * const index_size,
+        const size_type max_index_value) {
+    gdf_error err;
+    //Get array of indices that do not appear in r_index_ptr
+    thrust::device_vector<index_type> unmatched_indices =
+        create_missing_indices(
+                *r_index_ptr, max_index_value, *index_size);
+    CUDA_CHECK_LAST()
+
+    //Expand l_index_ptr and r_index_ptr if necessary
+    size_type mismatch_index_size = unmatched_indices.size();
+    size_type l_index_capacity = *index_capacity;
+    size_type r_index_capacity = *index_capacity;
+    err = expand_buffer(l_index_ptr, &l_index_capacity, *index_size, mismatch_index_size);
+    if (GDF_SUCCESS != err) return err;
+    err = expand_buffer(r_index_ptr, &r_index_capacity, *index_size, mismatch_index_size);
+    if (GDF_SUCCESS != err) return err;
+
+    //Copy JoinNoneValue to l_index_ptr to denote that a match does not exist on the left
+    thrust::fill(
+            thrust::device,
+            *l_index_ptr + *index_size,
+            *l_index_ptr + *index_size + mismatch_index_size,
+            JoinNoneValue);
+
+    //Copy unmatched indices to the r_index_ptr
+    thrust::copy(thrust::device,
```

Comment:
Be consistent with argument alignment. Notice difference from this line and line 165.